implementations/utils/gpu.py

```diff
@@ -0,0 +1,394 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities to create, delete, and SSH with GPUs."""
+
+from __future__ import annotations
+
+
+from absl import logging
+import airflow
+from airflow.decorators import task
+from apis import gcp_config, test_config
+import fabric
+from google.api_core.extended_operation import ExtendedOperation
+from google.cloud import compute_v1
+from implementations.utils import ssh
+import io
+import paramiko
+import re
+import sys
+from typing import Iterable, Any
+import uuid
+
+
+def get_image_from_family(project: str, family: str) -> compute_v1.Image:
+  """
+  Retrieve the newest image that is part of a given family in a project.
+
+  Args:
+    project: project ID or project number of the Cloud project you want to get image from.
+    family: name of the image family you want to get image from.
+
+  Returns:
+    An Image object.
+  """
+  image_client = compute_v1.ImagesClient()
+  # List of public operating system (OS) images: https://cloud.google.com/compute/docs/images/os-details
+  newest_image = image_client.get_from_family(project=project, family=family)
+  return newest_image
+
+
+def disk_from_image(
+    disk_type: str,
+    disk_size_gb: int,
+    boot: bool,
+    source_image: str,
+    auto_delete: bool = True,
+) -> compute_v1.AttachedDisk:
+  """
+  Create an AttachedDisk object to be used in VM instance creation. Uses an image as the
+  source for the new disk.
+
+  Args:
+    disk_type: the type of disk you want to create. This value uses the following format:
+        "zones/{zone}/diskTypes/(pd-standard|pd-ssd|pd-balanced|pd-extreme)".
+        For example: "zones/us-west3-b/diskTypes/pd-ssd"
+    disk_size_gb: size of the new disk in gigabytes
+    boot: boolean flag indicating whether this disk should be used as a boot disk of an instance
+    source_image: source image to use when creating this disk. You must have read access to this disk. This can be one
+        of the publicly available images or an image from one of your projects.
+        This value uses the following format: "projects/{project_name}/global/images/{image_name}"
+    auto_delete: boolean flag indicating whether this disk should be deleted with the VM that uses it
+
+  Returns:
+    AttachedDisk object configured to be created using the specified image.
+  """
+  boot_disk = compute_v1.AttachedDisk()
+  initialize_params = compute_v1.AttachedDiskInitializeParams()
+  initialize_params.source_image = source_image
+  initialize_params.disk_size_gb = disk_size_gb
+  initialize_params.disk_type = disk_type
+  boot_disk.initialize_params = initialize_params
+  # Remember to set auto_delete to True if you want the disk to be deleted when you delete
+  # your VM instance.
+  boot_disk.auto_delete = auto_delete
+  boot_disk.boot = boot
+  return boot_disk
+
+
+def create_instance(
```

Comment:
What's the benefit of separating this from `create_resource`? IMO it's more readable to inline the parts you need there, like `compute_v1.Instance(name=instance_name, ...)`

Reply:
My thinking is that `create_resource` is the function that takes input related to our automation setup like customized metadata, while `create_instance` is a more general api for creating a VM.

Reply:
I got what you mean here. Will remove the `wait_for_extended_operation` and use sensor instead for resource creation in a separate PR. I have updated the solution to resource deletion.